web3_proxy/src/frontend/users/referral.rs

```diff
@@ -70,3 +75,132 @@ pub async fn user_referral_link_get(
     let response = (status_code, Json(response_json)).into_response();
     Ok(response)
 }
+
+#[debug_handler]
+pub async fn user_used_referral_stats(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    TypedHeader(Authorization(bearer)): TypedHeader<Authorization<Bearer>>,
+    Query(_params): Query<HashMap<String, String>>,
+) -> Web3ProxyResponse {
+    // First get the bearer token and check if the user is logged in
+    let (user, _semaphore) = app.bearer_is_authorized(bearer).await?;
+
+    let db_replica = app
+        .db_replica()
+        .context("getting replica db for user's revert logs")?;
+
+    // Get all referral records associated with this user
+    let referrals = referee::Entity::find()
+        .filter(referee::Column::UserId.eq(user.id))
+        .find_also_related(referrer::Entity)
+        .all(db_replica.conn())
+        .await?;
+
+    // For each related referral person, find the corresponding user-address
+    #[derive(Debug, Serialize)]
+    struct Info {
+        credits_applied_for_referee: bool,
+        credits_applied_for_referrer: Decimal, // Probably make this a float instead (approximate anyways)
+        referral_start_date: DateTime,
+        used_referral_code: String,
+    };
+
+    let mut out: Vec<Info> = Vec::new();
+    for x in referrals.into_iter() {
+        let (referral_record, referrer_record) = (x.0, x.1.unwrap());
```

Comment:
is unwrap needed/safe here?

Reply:
each referral_record comes with a referrer, technically it should be (each referral code has a referrer). i can modify it to use context though

Reply:
changed it to context